File: dotnet/src/SimpleOrm/SnakeCaseNamingConvention.cs
using System.Text;

namespace SimpleOrm;

/// <summary>
/// The default convention: PascalCase/camelCase → snake_case. The exact algorithm is
/// part of the spec (ports must produce identical database names from their own
/// idioms), pinned by the test vectors in the reference implementation:
/// <c>UserId</c> → <c>user_id</c>, <c>UserID</c> → <c>user_id</c>,
/// <c>APIKey</c> → <c>api_key</c>, <c>HTMLParser</c> → <c>html_parser</c>,
/// <c>Address2</c> → <c>address2</c>, <c>Address2B</c> → <c>address2_b</c>.
/// </summary>
public sealed class SnakeCaseNamingConvention : INamingConvention
{
    /// <summary>Shared instance; the class is stateless.</summary>
    public static SnakeCaseNamingConvention Instance { get; } = new();

    public string ColumnName(string propertyName) => ToSnakeCase(propertyName);

    public string TableName(string typeName) => ToSnakeCase(typeName);

    public string IndexName(string tableName, IReadOnlyList<string> columnNames)
    {
        var builder = new StringBuilder("ix_").Append(tableName);
        foreach (var column in columnNames)
        {
            builder.Append('_').Append(column);
        }

        return builder.ToString();
    }

    /// <summary>
    /// An underscore is inserted before an upper-case letter that follows a
    /// lower-case letter or digit, or that starts the last word of an acronym run
    /// (an upper followed by a lower); everything is lower-cased.
    /// </summary>
    private static string ToSnakeCase(string name)
    {
        if (name.Length == 0)
        {
            return name;
        }

        var builder = new StringBuilder(name.Length + 4);
        for (var i = 0; i < name.Length; i++)
        {
            var c = name[i];
            if (char.IsUpper(c) && i > 0)
            {
                var previous = name[i - 1];
                var startsWordAfterLowerOrDigit = char.IsLower(previous) || char.IsDigit(previous);
                var endsAcronymRun = char.IsUpper(previous) && i + 1 < name.Length && char.IsLower(name[i + 1]);
                if (startsWordAfterLowerOrDigit || endsAcronymRun)
                {
                    builder.Append('_');
                }
            }

            builder.Append(char.ToLowerInvariant(c));
        }

        return builder.ToString();
    }
}

File: php/src/Naming/SnakeCaseNamingConvention.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Naming;

/**
 * The default convention (spec/metadata-model.md "Naming convention"): PascalCase
 * or camelCase to snake_case. The algorithm is part of the spec — every port must
 * produce identical database names from its own idiom — pinned by the normative
 * vectors there and mirrored from the C# reference's SnakeCaseNamingConvention.
 */
final class SnakeCaseNamingConvention implements NamingConvention
{
    public function toDatabase(string $name): string
    {
        return self::toSnakeCase($name);
    }

    /**
     * An underscore is inserted before an upper-case letter that follows a
     * lower-case letter or digit, or that starts the last word of an acronym run
     * (an upper followed by a lower); everything is lower-cased.
     */
    private static function toSnakeCase(string $name): string
    {
        $length = strlen($name);
        if ($length === 0) {
            return $name;
        }

        $result = '';
        for ($i = 0; $i < $length; $i++) {
            $c = $name[$i];
            if (self::isUpper($c) && $i > 0) {
                $previous = $name[$i - 1];
                $startsWordAfterLowerOrDigit = self::isLower($previous) || self::isDigit($previous);
                $endsAcronymRun = self::isUpper($previous) && $i + 1 < $length && self::isLower($name[$i + 1]);
                if ($startsWordAfterLowerOrDigit || $endsAcronymRun) {
                    $result .= '_';
                }
            }

            $result .= strtolower($c);
        }

        return $result;
    }

    private static function isUpper(string $c): bool
    {
        return $c >= 'A' && $c <= 'Z';
    }

    private static function isLower(string $c): bool
    {
        return $c >= 'a' && $c <= 'z';
    }

    private static function isDigit(string $c): bool
    {
        return $c >= '0' && $c <= '9';
    }
}
